Introduce ESLint to svelte sources

43081j opened the pull request:
Fixes #2951.

Are you interested in this?

Interestingly the PR guidelines/template states to remember to run `npm run lint`, but there wasn't such a script before..

Here's what I've done:

* Cleared most of `.eslintignore` as those files don't seem to exist anymore
* Added ESLint, the typescript parser/plugin and the import plugin (which your config already tried to extend)
* Disabled the rules which seem to go against your conventions (e.g. camel case identifiers)
* Added ESLint as an npm script for `test/` and `src/`

If you're interested in taking this PR on im happy to sort the mass of lint errors currently produced so we can get it merged. Most are just inconsistencies with whitespace or interface separator style (semicolon vs comma).

run1t commented:
I Just open an issue about this yesterday (#2958) 😄
I'd be happy to help if you need to !

43081j commented:
alright so tonight i fixed the many thousands of lint errors.

some by disabling rules (as svelte's conventions differ), some automatically (`--fix`) and some myself.

any rules I disabled were either because we had a large enough amount of code that it clearly didn't match convention or because typescript conflicts with it.

---

Here's a note to self and anyone who is interested:

* Odd indentation
	* `nodes/interfaces` L33
	* `render-dom/index` L79
	* `wrappers/EachBlock` L312
	* `wrappers/IfBlock` L213
	* `wrappers/InlineComponent` L234
	* `wrappers/Window` L142
	* `compiler/interfaces` L10 + L41
* Odd unicode
	* `compile/Component`
	* `compile/css/Stylesheet`
	* `compile/nodes/Element`
	* `nodes/shared/Expression`
	* `render-dom/index`
	* `wrappers/DebugTag`
	* `wrappers/Element/Attribute`
	* `wrappers/Element/Binding`
	* `wrappers/Framgnet`
	* `wrappers/InlineComponent/index`
	* `wrappers/Title`
	* `wrappers/shared/Wrapper`
	* `render-ssr/index`
	* `parse/state/tag`
	* `utils/html`

These are some oddities that happened after `eslint --fix`, the unicode especially is scattered around. I read through the entire diff and these are the places that looked out of place.

---

Also, I left the lint fixes in their own commit in case we want to discard them and keep the setup/config.

Rich-Harris commented on src/compiler/compile/render-dom/wrappers/Title.ts:
```diff
@@ -14,13 +14,13 @@ export default class TitleWrapper extends Wrapper {
 		node: Title,
-		strip_whitespace: boolean,
-		next_sibling: Wrapper
+		_strip_whitespace: boolean,
```
Is there a rule for controlling whether unused variables need a leading underscore? I'm in two minds about it...

43081j replied:
ESLint's rule is configurable such that those matching the specified pattern are allowed.

I originally set it to `^_`, but actually later enabled typescript's built in (stricter) unused checks at compile time and turned the eslint rule off.

TypeScript has built in support for ignoring `_` prefixed locals/parameters. I wouldn't go with another convention because this is built into typescript its self.

Rich-Harris commented:
One other thing — a lot of the changes are reordering functions etc. Is that necessary? It renders the git history slightly less useful...

43081j commented:
I do realise that and it comes down to preference really.

We can either turn the rule off and leave it as it was, or turn it on and reorder them.

I noticed its pretty inconsistent, so wasn't sure which would be preferred. Many files have functions at the start, and at the end. Many have them at the end only and some have them at the start only.

43081j commented on src/compiler/compile/render-dom/wrappers/IfBlock.ts:
```diff
@@ -429,7 +429,7 @@ export default class IfBlockWrapper extends Wrapper {
 			`;
 
-		// no `p()` here — we don't want to update outroing nodes,
+		// no `p()` here ÃÂ¢ÃÂÃÂ we don't want to update outroing nodes,
```
its a unicode dash of some sort, `eslint --fix` doesn't seem to have dealt with it well so rewrote it with broken unicode. i plan on reverting these lines as i find them (though a better solution might be to just use a regular hyphen...)

Rich-Harris replied:
> a better solution might be to just use a regular hyphen

touché

Conduitry commented:
I'd vote for not changing the order that stuff is in the files. It might have been nice if there had been a more consistent ordering from the beginning, but I think this is actually of limited value given that editors have a 'jump to definition' command.

43081j commented on src/compiler/compile/render-dom/wrappers/IfBlock.ts:
```diff
@@ -213,8 +213,8 @@ export default class IfBlockWrapper extends Wrapper {
 				${this.branches
-					.map(({ condition, block }) => `${condition ? `if (${condition}) ` : ''}return ${block.name};`)
-					.join('\n')}
+		.map(({ condition, block }) => `${condition ? `if (${condition}) ` : ''}return ${block.name};`)
```
eslint has some real trouble with correcting whitespace in template expressions. i'd suggest discarding these whitespace changes and using a formatter instead, or at least discarding the bad ones and setting the rule to `warn` instead

43081j replied:
alternatively we could have less logic inline in template expressions and extract it out into local variables, so the template becomes as simple as `${branchesMap}` for example.

Rich-Harris replied:
My experience of Prettier hasn't been all that great either to be honest — I tend to think that this sort of formatting is best left to humans

43081j commented:
thats fair enough.

i agree.

i reverted ordering changes and reverted whitespace changes in template expressions. i also changed the indent rule to a warning instead (as it will error on those still)

43081j commented:
i've rolled back some more unicode mess and i re-enabled the indent rule as `error` level (ignored `TemplateLiteral` direct child nodes and ignored some specific blocks of code).

weirdly the tests no longer pass, not too sure what happened there as the diff is the same as before but cleaner.

Rich-Harris commented:
thank you!
